### ISSUE TITLE
Monitor crashing after a few seconds

### ISSUE THREAD

**Issue by threalboss** [sam210723/goesrecv-monitor#2]
HI,
I just got this and whenever i connect to my pi the monitor will work for 1 second then crash. Any help?

**Comment by robman501a** [comment]
Hi, I have the same problem. I have Windows 10 x64 and .NET 4.6.1 is saying it's installed.

**Comment by sam210723** [comment]
Reported [via Twitter](https://twitter.com/NautilusCL/status/1189700756137545731) as well. Investigating...

![Twitter Image 1](https://user-images.githubusercontent.com/5117093/68540640-2adef080-03e1-11ea-9cdb-13e3e785baba.png)
![Twitter Image 2](https://user-images.githubusercontent.com/5117093/68540641-30d4d180-03e1-11ea-813a-999b3fa9ddcc.png)
![Twitter Image 3](https://user-images.githubusercontent.com/5117093/68540642-34685880-03e1-11ea-8239-b7ae818e6b9e.png)

**Comment by sam210723** [comment]
Verbose logging has been added in the v1.1 branch. It creates ``log.txt`` in the same directory as the executable.

I can send a compiled version to anyone experiencing this crash to try narrow down the cause.

**Comment by sam210723** [comment]
Cleaned up JSON string processing in 2ca951be and fd6be9a. Seems to be more reliable based on log output. Waiting for user on Twitter to confirm the issue has been fixed.

**Comment by threalboss** [comment]
Should I re-download the 1.2 version and try and see if it is fixed?

**Comment by sam210723** [comment]
That'd be great if you could. There's a pre-built debug version available at https://vksdr.com/downloads/goesrecv-monitor-debug.zip.

User on Twitter confirmed there is still an issue, however it's worth trying on another system. Log file can be sent to sam at vksdr dot com.

Are you also able to send me your goesrecv executable? I'd like to run it locally.

**Comment by robman501a** [comment]
Hi, I'm having the same issue and I am using v1.1. 
This was the only way that I could get it running for about 30 seconds.
![image](https://user-images.githubusercontent.com/30871515/80927776-24b54580-8d6e-11ea-9e6c-51a072797dc4.png)

and when it crashed, I found this in the event logs.
I'm running Windows 10 x64 and the latest version of .NET.
```
Application: goesrecv-monitor.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.ArgumentOutOfRangeException
   at System.ThrowHelper.ThrowArgumentOutOfRangeException(System.ExceptionArgument, System.ExceptionResource)
   at goesrecv_monitor.Stats.DecoderLoop()
   at System.Threading.ThreadHelper.ThreadStart_Context(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.ThreadHelper.ThreadStart()

```

**Comment by sam210723** [comment]
Hi, thanks for the event log info. Could you please [follow the instructions here](https://github.com/sam210723/goesrecv-monitor/tree/master/README.md#debug-logs) to enable logging, then try running the program again. It will create ``log.txt`` with extra info I can use for debugging.

Email this file to ``sam at vksdr dot com`` and I'll try find a fix for this crash!

**Comment by robman501a** [comment]
Okay, I sent you all the data I could find.

**Comment by Adrinus** [comment]
Same issue here:
`[07/05/2020 17:25:39.094] [DECODER]   Error parsing JSON: Newtonsoft.Json.JsonReaderException: Unterminated string. Expected delimiter: ". Path 'viterbi_errors', line 1, position 96.
   at Newtonsoft.Json.JsonTextReader.ReadStringIntoBuffer(Char quote)
   at Newtonsoft.Json.JsonTextReader.ParseProperty()
   at Newtonsoft.Json.JsonTextReader.ParseObject()
   at Newtonsoft.Json.JsonTextReader.Read()
   at Newtonsoft.Json.Linq.JContainer.ReadContentFrom(JsonReader r, JsonLoadSettings settings)
   at Newtonsoft.Json.Linq.JContainer.ReadTokenFrom(JsonReader reader, JsonLoadSettings options)
   at Newtonsoft.Json.Linq.JObject.Load(JsonReader reader, JsonLoadSettings settings)
   at Newtonsoft.Json.Linq.JObject.Parse(String json, JsonLoadSettings settings)
   at goesrecv_monitor.Stats.DecoderLoop()
                                      {"timestamp": "2020-05-07T21:25:38.162Z","skipped_symbols": 16417,"viterbi_errors": 2315,"reed_s`

And

`[07/05/2020 17:25:39.094] [DECODER]   No JSON object found
[07/05/2020 17:25:39.095] [DECODER]   Error parsing JSON: Newtonsoft.Json.JsonReaderException: Unterminated string. Expected delimiter: ". Path 'skipped_symbols', line 1, position 72.
   at Newtonsoft.Json.JsonTextReader.ReadStringIntoBuffer(Char quote)
   at Newtonsoft.Json.JsonTextReader.ParseProperty()
   at Newtonsoft.Json.JsonTextReader.ParseObject()
   at Newtonsoft.Json.JsonTextReader.Read()
   at Newtonsoft.Json.Linq.JContainer.ReadContentFrom(JsonReader r, JsonLoadSettings settings)
   at Newtonsoft.Json.Linq.JContainer.ReadTokenFrom(JsonReader reader, JsonLoadSettings options)
   at Newtonsoft.Json.Linq.JObject.Load(JsonReader reader, JsonLoadSettings settings)
   at Newtonsoft.Json.Linq.JObject.Parse(String json, JsonLoadSettings settings)
   at goesrecv_monitor.Stats.DecoderLoop()
                                      {"timestamp": "2020-05-07T21:25:38.358Z","skipped_symbols": 4143,"viterb`

These are the only errors the log throws.

**Comment by sam210723** [comment]
Ok interesting. Would you be able to send the full log to the email address a few comments back?

**Comment by Adrinus** [comment]
Sent. :)

**Comment by sam210723** [comment]
Hi all,

I'll be releasing ``v1.2`` tomorrow which receives data from goesrecv much more reliably. Each JSON string sent by goesrecv has an 8 byte header before it which indicates the length of the JSON string following the header. goesrecv-monitor will now read that length value first then receive the correct number of bytes for that JSON string. These bytes are then converted to ASCII and parsed like normal. Before, this header was being ignored then parsed as ASCII which is what created erroneous characters and null bytes in the JSON strings.

![Socket bytes](https://user-images.githubusercontent.com/5117093/85223958-22847480-b40a-11ea-932d-a618d86a8adb.png)

This release also includes memory usage optimisations so the .NET garbage collector is called far less often.. In previous versions it was often called a few seconds after starting then multiple times per minute after that. Garbage collection only runs every few minutes now.

![Memory usage snapshots](https://user-images.githubusercontent.com/5117093/85226427-beb67780-b41a-11ea-893f-2fbccd7442d9.png)
![Memory usage graph](https://user-images.githubusercontent.com/5117093/85226444-d3930b00-b41a-11ea-9356-32c289f57cf8.png)

It'd be great if someone can try this version out as I am still unable to recreate the fault here.

**Comment by sam210723** [comment]
Released [v1.2](https://github.com/sam210723/goesrecv-monitor/releases/tag/v1.2). Binary ZIP attached to release.
